Avoid creating folder for ttnt files

Genki-S opened the pull request:
Implementation of #21 

I changed the saving mechanism from saving to separated file under `.ttnt` directory to saving to a single file named `.ttnt`.

Since I did not like to touch the same file from many places across the application, I made these changes:

- created `TTNT::Storage` class, which is responsible for IO with the `.ttnt` file
- created `TTNT::Metadata` class, which uses `TTNT::Storage` to store and retrieve metadata
- made `TTNT::TestToCodeMapping` use `TTNT::Storage` to save mappings to the `.ttnt` file

`TTNT::Storage` can have `sections`, which corresponds to the top level key of JSON object in `.ttnt` file (i.e. if `.ttnt` has `{ 'meta': {...}, 'mapping': {...} }`, there are 'meta' and 'mapping' sections).

And since we should be able to read file contents from a given commit on the git history (e.g. when a user ran `rake ttnt:test:anchor` and `.ttnt` file on current working tree is changed and the user runs `rake ttnt:test:run`, the mapping should be read from `.ttnt` file from the commit `HEAD` points to), I made it possible here: https://github.com/Genki-S/ttnt/commit/c2909fdcc93032760837503dc7e91b6c9c56dade
(which is actually a generalization of the logic [here](https://github.com/Genki-S/ttnt/pull/25/files#diff-c2993c7789d2d14196b4b1b184845358L76))

@robin850 could you review the code please?

(I have some refactoring other than these changes, so I will make another PR for those once this gets merged)

robin850 commented on test/metadata_test.rb:
```diff
@@ -0,0 +1,27 @@
+class MetaDataTest < TTNT::TestCase
+  def setup
+    @storage_file = "#{@repo.workdir}/.ttnt"
+    FileUtils.rm_rf(@storage_file)
```
You could simply use `File.delete` here ; this seems less overkill and more safe anyway.

Genki-S replied:
That's true, thank you! Fixed: https://github.com/Genki-S/ttnt/commit/1ec9320e6d6dff2aaa6e1d42fe42807477255e2c

robin850 commented on test/storage_test.rb:
```diff
@@ -0,0 +1,40 @@
+class StorageTest < TTNT::TestCase
+  def setup
+    @storage_file = "#{@repo.workdir}/.ttnt"
+    FileUtils.rm_rf(@storage_file)
```
Ditto regarding `File.delete`.

Genki-S replied:
Fixed in this too: https://github.com/Genki-S/ttnt/commit/1ec9320e6d6dff2aaa6e1d42fe42807477255e2c

robin850 commented on test/test_to_code_mapping_test.rb:
```diff
@@ -5,7 +5,7 @@ class TestToCodeMappingTest < TTNT::TestCase
   def setup
     # clean up generated .ttnt files
     FileUtils.rm_rf("#{@repo.workdir}/.ttnt")
```
Ditto regarding `File.delete`.

Genki-S replied:
Fixed in this too: https://github.com/Genki-S/ttnt/commit/1ec9320e6d6dff2aaa6e1d42fe42807477255e2c

robin850 commented on lib/ttnt/storage.rb:
```diff
@@ -0,0 +1,67 @@
+      if @sha
+        @repo.lookup(@repo.lookup(@sha).tree['.ttnt'][:oid]).content
+      else
+        if File.exist?(filename)
```
You could rewrite is as a ternary like:

~~~ruby
File.exist?(filename) ? File.read(filename) : ""
~~~

Genki-S replied:
That's cool, thanks! Changed here: https://github.com/Genki-S/ttnt/commit/a27c40b141976cc2c2cce693d2053aa1cab73417

robin850 commented on lib/ttnt/storage.rb:
```diff
@@ -0,0 +1,67 @@
+      File.open(filename, File::RDWR|File::CREAT, 0644) do |f|
+        f.flock(File::LOCK_EX)
+        str = f.read
+        data = if str.length > 0 then JSON.parse(str) else {} end
```
This one as well could be rewritten as a ternary:

~~~ruby
data = str.length > 0 ? JSON.parse(str) : {}
~~~

Genki-S replied:
I don't know why, but it seems I thought there are no ternary operator in Ruby :sweat: Fixed: https://github.com/Genki-S/ttnt/commit/a27c40b141976cc2c2cce693d2053aa1cab73417

robin850 commented on lib/ttnt/test_to_code_mapping.rb:
```diff
@@ -9,10 +10,15 @@ module TTNT
     # @param repo [Rugged::Reposiotry] repository to save test-to-code mapping
     #   (only repo.workdir is used to determine where to save the mapping file)
-    def initialize(repo)
+    # @param sha [String] sha of commit from which mapping is read from.
```
There's an extra "from" here. :-p

Genki-S replied:
OMG :scream: Thank you for catching! Fixed: https://github.com/Genki-S/ttnt/commit/0e8d34d13ab08d6eb208b031cef67b8940a3ba48

robin850 commented on lib/ttnt/storage.rb:
```diff
@@ -0,0 +1,67 @@
+module TTNT
+  # A storage to store TTNT data such as test-to-code mapping and metadata.
```
"storage" seems redundant here. What do you think about "A utility class to store TTNT data ..." ?

Genki-S replied:
Sounds much better! :+1: Changed: https://github.com/Genki-S/ttnt/commit/1af60bab207ba561f6f8b15091943f7a6286f621

robin850 commented on test/storage_test.rb:
```diff
@@ -0,0 +1,40 @@
+    new_data = { 'c' => 3 }
+    @storage.write!(@section, new_data) # write to a file in working tree
+    history_storage = TTNT::Storage.new(@repo, sha)
+    assert_equal @data, history_storage.read(@section)
```
Don't you think that something like:

~~~ruby
assert !history_storage.read(@section).key?('c')
~~~

show more the intent of the test ? I've no strong feeling about this one though.

Genki-S replied:
That's exactly what I wanted to test! :bulb:  Thank you for letting me know how to express my intention in test code, this is really amazing :smile: Fixed: https://github.com/Genki-S/ttnt/commit/9341f7f7cd48062962995a263b00ac491e7cecac

robin850 commented:
Nice work Genki! :-)

A few last comments:

* What do you think about renaming `TTNT::MetaData`'s `#get` and `#set` to `#[]=` and `#[]` ? This seems more idiomatic and relevant as we are dealing with JSON/Hash objects.
* You could actually define a default value of `nil` for the `sha` parameter for `MetaData#initialize` and `TestToCodeMapping#initialize`. This will be consistent with `Storage`'s.

Genki-S commented:
> What do you think about renaming TTNT::MetaData's #get and #set to #[]= and #[] ? This seems more idiomatic and relevant as we are dealing with JSON/Hash objects

Seems much better! I'm not yet used to define `[]` `[]=` methods :disappointed: Thank you for mentioning! Changed: https://github.com/Genki-S/ttnt/commit/a93f0676536c78aea22c4e701a3aa8aec63547a1

> You could actually define a default value of nil for the sha parameter for MetaData#initialize and TestToCodeMapping#initialize. This will be consistent with Storage's.

That's true, thank you! Fixed: https://github.com/Genki-S/ttnt/commit/cebc769ecd37a8ba74d5878a49954fcdf8708bf8
